### PR TITLE
ARROW-2483: [Rust] use bit-packing for boolean vectors

### PR DESCRIPTION
**Pull request by sunchao**

### CODE REVIEW

**Comment by sunchao** [comment]
@kszucs @paddyhoran @andygrove @crepererum : could you review this? Thanks!

File: rust/src/array.rs
Comment (kszucs) [inline rust/src/array.rs]: type `an boolean`

File: rust/src/array.rs
Comment (kszucs) [inline rust/src/array.rs]: Just a note, that `pa` naming confuses me a little, because the python impl uses `import pyarrow as pa` :)

**Review by kszucs — APPROVED** [review]
Nice change! I'll let others to review too before merging.

File: rust/src/array.rs
Comment (paddyhoran) [inline rust/src/array.rs]: Shouldn't this be something like "Note this does not take the offset of the underlying `ArrayData` into account"

**Review by paddyhoran — APPROVED** [review]
+1 LGTM

We probably don't need `Bitmap` with this in place, right?

File: rust/src/array.rs
Comment (sunchao) [reply rust/src/array.rs]: Yes offset technically is a property of `ArrayData`. However it is also (indirectly) a property of a `Array` since `ArrayData` encapsulates the common properties for all arrays. So, IMO this comment should be fine.

**Comment by sunchao** [comment]
Thanks @kszucs and @paddyhoran for the review. 

> We probably don't need Bitmap with this in place, right?

`Bitmap` is a thin wrapper around the `get_bit` operations, so it can be removed easily if we decide to. The `BooleanArray` is an array type which contains both a null vector and a value vector. The common of the two is that they all use the same bit operations.

File: rust/src/array.rs
Comment (paddyhoran) [reply rust/src/array.rs]: ok, sounds good

**Review by crepererum — CHANGES_REQUESTED** [review]

**Comment by andygrove** [comment]
@sunchao I'll be happy to merge this once the two issues raised by @crepererum are addressed

File: rust/src/array.rs
Comment (paddyhoran) [reply rust/src/array.rs]: Isn't this the one situation where using a wildcard is acceptable (i.e. bringing all the contents of the module into scope so it can be tested)?

Genuinely asking the question on what is best practice rather than disagreeing :)

Comment (crepererum) [reply rust/src/array.rs]: Ah, you're right. It's a very common pattern for tests. Sorry for the noise. 🙂

File: rust/src/util/bit_util.rs
Comment (sunchao) [reply rust/src/util/bit_util.rs]: I basically followed the [C++ version](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit-util.h#L318) for the implementation, but I think you raised some very good points here. Replaced the ops with bit ops, and also added bound check in the caller site. 

For the `i64` type issue, I think it is valid to change to `usize` but that's a broader change than this PR. I'll do it separately.

File: rust/src/util/bit_util.rs
Comment (crepererum) [inline rust/src/util/bit_util.rs]: Same as above: what happens if i is negative?

File: rust/src/util/bit_util.rs
Comment (crepererum) [reply rust/src/util/bit_util.rs]: Why can i not be an unsigned integer?

**Review by crepererum — CHANGES_REQUESTED** [review]

File: rust/src/util/bit_util.rs
Comment (sunchao) [reply rust/src/util/bit_util.rs]: Done.

File: rust/src/util/bit_util.rs
Comment (sunchao) [reply rust/src/util/bit_util.rs]: The caller is responsible for making sure that `i` stays within bounds. As I mentioned in the last comment, will do a separate PR to change the `i64` to `usize`. Same below.

File: rust/src/util/bit_util.rs
Comment (sunchao) [reply rust/src/util/bit_util.rs]: Same above. Will change this to `usize` in a separate PR (since it's a broader change).

**Comment by crepererum** [comment]
+1 good to go!
Thank you @sunchao for the work!

**Comment by sunchao** [comment]
Thanks @crepererum for the review!

**Comment by kszucs** [comment]
Thanks everyone! Merging

**Pull request closed without merging**
